[BUG] API key is required for vllm

**Describe the bug**
When using sim with VLLM_BASE_URL, sim can see and list all the model hosted behind but throw an 
`API key is required for vllm vllm/qwen-3` error message on use. whether the VLLM_API_KEY variable is set or not. 

vLLM is reachable on all other side app, without any api key. 

**To Reproduce**
Steps to reproduce the behavior:

1. Start an [vLLM](https://github.com/vllm-project/vllm) instance
2. Deploy Sim using helm-chart (based on this [values](https://github.com/simstudioai/sim/blob/main/helm/sim/examples/values-aws.yaml) for me )
3. Setup VLLM_BASE_URL under `app.env` key 
```
app:
  env: 
    ...
    VLLM_BASE_URL: "https:/vllm.exemple.com"
    ...
```
4. Open an agent in editor and select our vLLM model
5. Click on 'Run' to test connectivity 

**Expected behavior**
Can use vLLM hosted model

**Screenshots**

<img width="202" height="151" alt="Image" src="https://github.com/user-attachments/assets/17761400-947c-43e8-a5e1-f26981dfb5e8" />

<img width="1880" height="186" alt="Image" src="https://github.com/user-attachments/assets/ae683aef-5824-4bec-bcf6-fb2e7bb39553" />

**Additional context**

Chart version: v0.5.29 
App version: v0.5.29

Full log on app side: 
```
[2025-12-15T13:33:40.083Z] [ERROR] [AgentBlockHandler] Failed to get API key: {"provider":"vllm","model":"vllm/qwen-3","error":"API key is required for vllm vllm/qwen-3","hasProvidedApiKey":false}
[2025-12-15T13:33:40.083Z] [ERROR] [BlockExecutor] Block execution failed {"blockId":"f83f3c4e-b453-476a-b963-9f21db2211e9","blockType":"agent","error":"API key is required for vllm vllm/qwen-3"}
[2025-12-15T13:33:40.084Z] [ERROR] [ExecutionEngine] Node execution failed {"nodeId":"f83f3c4e-b453-476a-b963-9f21db2211e9","error":"[agent] Agent 1: API key is required for vllm vllm/qwen-3"}
[2025-12-15T13:33:40.084Z] [ERROR] [ExecutionEngine] Execution failed {"error":"[agent] Agent 1: API key is required for vllm vllm/qwen-3"}
[2025-12-15T13:33:40.084Z] [ERROR] [ExecutionCore] [bb20a523] Execution failed: {"message":"[agent] Agent 1: API key is required for vllm vllm/qwen-3","blockId":"f83f3c4e-b453-476a-b963-9f21db2211e9","blockName":"Agent 1","blockType":"agent","workflowId":"46640e50-ee47-4e34-a432-2062e44949c7","timestamp":"2025-12-15T13:33:40.084Z","nodeId":"f83f3c4e-b453-476a-b963-9f21db2211e9","executionTime":0,"executionResult":{"success":false,"output":{},"error":"[agent] Agent 1: API key is required for vllm vllm/qwen-3","logs":[{"blockId":"ed396b24-8488-4c2c-b8d1-2fcecaa3b065","blockName":"Start","blockType":"start_trigger","startedAt":"2025-12-15T13:33:40.082Z","endedAt":"2025-12-15T13:33:40.082Z","durationMs":0,"success":true,"input":{"inputFormat":[{"id":"43256295-8ba7-431b-a1bc-d422916a0dd9","name":"username","type":"string","value":"foo","collapsed":false}],"triggerMode":true},"output":{"username":"foo"}},{"blockId":"f83f3c4e-b453-476a-b963-9f21db2211e9","blockName":"Agent 1","blockType":"agent","startedAt":"2025-12-15T13:33:40.083Z","endedAt":"2025-12-15T13:33:40.083Z","durationMs":0,"success":false,"input":{"model":"vllm/qwen-3","tools":[],"apiKey":null,"messages":[{"role":"user","content":"Say hello to foo."}],"verbosity":"medium","memoryType":"none","temperature":0.3,"azureEndpoint":null,"conversationId":"session-foo","responseFormat":null,"azureApiVersion":null,"reasoningEffort":"medium","slidingWindowSize":null,"slidingWindowTokens":null},"error":"API key is required for vllm vllm/qwen-3"}],"metadata":{"requestId":"bb20a523","executionId":"444cb113-7b72-474f-8602-d301e227c1a3","workflowId":"46640e50-ee47-4e34-a432-2062e44949c7","workspaceId":"74296bf1-9476-488a-9771-ebe4ab1c24ab","userId":"hugvHtQNLfMfYOL6unmxRRBj90QFBAUP","sessionUserId":"hugvHtQNLfMfYOL6unmxRRBj90QFBAUP","workflowUserId":"hugvHtQNLfMfYOL6unmxRRBj90QFBAUP","triggerType":"manual","useDraftState":true,"startTime":"2025-12-15T13:33:40.082Z","isClientSession":true,"workflowStateOverride":{"blocks":{"ed396b24-8488-4c2c-b8d1-2fcecaa3b065":{"id":"ed396b24-8488-4c2c-b8d1-2fcecaa3b065","type":"start_trigger","name":"Start","position":{"x":-797.6923076923077,"y":-672.3076923076922},"enabled":true,"horizontalHandles":true,"advancedMode":false,"triggerMode":false,"height":0,"subBlocks":{"inputFormat":{"id":"inputFormat","type":"input-format","value":[{"id":"43256295-8ba7-431b-a1bc-d422916a0dd9","name":"username","type":"string","value":"Brice","collapsed":false}]}},"outputs":{"files":{"type":"files","description":"User uploaded files"},"input":{"type":"string","description":"Primary user input or message"},"conversationId":{"type":"string","description":"Conversation thread identifier"}},"data":{}},"f83f3c4e-b453-476a-b963-9f21db2211e9":{"id":"f83f3c4e-b453-476a-b963-9f21db2211e9","type":"agent","name":"Agent 1","position":{"x":-498.84615384615375,"y":-756.153846153846},"enabled":true,"horizontalHandles":true,"advancedMode":false,"triggerMode":false,"height":0,"subBlocks":{"model":{"id":"model","type":"combobox","value":"vllm/qwen-3"},"tools":{"id":"tools","type":"tool-input","value":[]},"apiKey":{"id":"apiKey","type":"short-input","value":null},"messages":{"id":"messages","type":"messages-input","value":[{"role":"user","content":"Say hello to <start.username>."}]},"verbosity":{"id":"verbosity","type":"dropdown","value":"medium"},"memoryType":{"id":"memoryType","type":"dropdown","value":"none"},"temperature":{"id":"temperature","type":"slider","value":0.3},"azureEndpoint":{"id":"azureEndpoint","type":"short-input","value":null},"conversationId":{"id":"conversationId","type":"short-input","value":"session-foo"},"responseFormat":{"id":"responseFormat","type":"code","value":null},"azureApiVersion":{"id":"azureApiVersion","type":"short-input","value":null},"reasoningEffort":{"id":"reasoningEffort","type":"dropdown","value":"medium"},"slidingWindowSize":{"id":"slidingWindowSize","type":"short-input","value":null},"slidingWindowTokens":{"id":"slidingWindowTokens","type":"short-input","value":null}},"outputs":{"model":{"type":"string","description":"Model used for generation"},"tokens":{"type":"any","description":"Token usage statistics"},"content":{"type":"string","description":"Generated response content"},"toolCalls":{"type":"any","description":"Tool calls made"}},"data":{}}},"edges":[{"id":"9b4ef229-fbe9-4bb9-be1c-93756d220f22","source":"ed396b24-8488-4c2c-b8d1-2fcecaa3b065","target":"f83f3c4e-b453-476a-b963-9f21db2211e9","sourceHandle":"source","targetHandle":"target","type":"default","data":{}}],"loops":{},"parallels":{}},"duration":2,"endTime":"2025-12-15T13:33:40.084Z"}}}
Error: [agent] Agent 1: API key is required for vllm vllm/qwen-3
    at $ (.next/server/chunks/apps_sim_lib_workflows_executor_human-in-the-loop-manager_ts_98617cee._.js:1:15802)
    at handleBlockError (.next/server/chunks/apps_sim_lib_workflows_executor_human-in-the-loop-manager_ts_98617cee._.js:1:18370)
    at processTicksAndRejections (null) {
  blockId: 'f83f3c4e-b453-476a-b963-9f21db2211e9',
  blockName: 'Agent 1',
  blockType: 'agent',
  workflowId: '46640e50-ee47-4e34-a432-2062e44949c7',
  timestamp: '2025-12-15T13:33:40.084Z',
  nodeId: 'f83f3c4e-b453-476a-b963-9f21db2211e9',
  executionTime: 0,
  executionResult: [Object]
}
 ⨯ unhandledRejection:  Error: [agent] Agent 1: API key is required for vllm vllm/qwen-3
    at $ (.next/server/chunks/apps_sim_lib_workflows_executor_human-in-the-loop-manager_ts_98617cee._.js:1:15802)
    at handleBlockError (.next/server/chunks/apps_sim_lib_workflows_executor_human-in-the-loop-manager_ts_98617cee._.js:1:18370)
    at processTicksAndRejections (null) {
  blockId: 'f83f3c4e-b453-476a-b963-9f21db2211e9',
  blockName: 'Agent 1',
  blockType: 'agent',
  workflowId: '46640e50-ee47-4e34-a432-2062e44949c7',
  timestamp: '2025-12-15T13:33:40.084Z',
  nodeId: 'f83f3c4e-b453-476a-b963-9f21db2211e9',
  executionTime: 0,
  executionResult: [Object]
}
```

--- a/apps/sim/app/api/providers/ollama/models/route.ts
+++ b/apps/sim/app/api/providers/ollama/models/route.ts
@@ -45,7 +45,6 @@ export async function GET(request: NextRequest) {
       host: OLLAMA_HOST,
     })
 
-    // Return empty array instead of error to avoid breaking the UI
     return NextResponse.json({ models: [] })
   }
 }

--- a/apps/sim/app/api/providers/vllm/models/route.ts
+++ b/apps/sim/app/api/providers/vllm/models/route.ts
@@ -20,10 +20,16 @@ export async function GET(request: NextRequest) {
       baseUrl,
     })
 
+    const headers: Record<string, string> = {
+      'Content-Type': 'application/json',
+    }
+
+    if (env.VLLM_API_KEY) {
+      headers.Authorization = `Bearer ${env.VLLM_API_KEY}`
+    }
+
     const response = await fetch(`${baseUrl}/v1/models`, {
-      headers: {
-        'Content-Type': 'application/json',
-      },
+      headers,
       next: { revalidate: 60 },
     })
 
@@ -50,7 +56,6 @@ export async function GET(request: NextRequest) {
       baseUrl,
     })
 
-    // Return empty array instead of error to avoid breaking the UI
     return NextResponse.json({ models: [] })
   }
 }

--- a/apps/sim/providers/utils.test.ts
+++ b/apps/sim/providers/utils.test.ts
@@ -53,19 +53,20 @@ describe('getApiKey', () => {
     module.require = originalRequire
   })
 
-  it('should return user-provided key when not in hosted environment', () => {
+  it.concurrent('should return user-provided key when not in hosted environment', () => {
     isHostedSpy.mockReturnValue(false)
 
-    // For OpenAI
     const key1 = getApiKey('openai', 'gpt-4', 'user-key-openai')
     expect(key1).toBe('user-key-openai')
 
-    // For Anthropic
     const key2 = getApiKey('anthropic', 'claude-3', 'user-key-anthropic')
     expect(key2).toBe('user-key-anthropic')
+
+    const key3 = getApiKey('google', 'gemini-2.5-flash', 'user-key-google')
+    expect(key3).toBe('user-key-google')
   })
 
-  it('should throw error if no key provided in non-hosted environment', () => {
+  it.concurrent('should throw error if no key provided in non-hosted environment', () => {
     isHostedSpy.mockReturnValue(false)
 
     expect(() => getApiKey('openai', 'gpt-4')).toThrow('API key is required for openai gpt-4')
@@ -74,63 +75,87 @@ describe('getApiKey', () => {
     )
   })
 
-  it('should fall back to user key in hosted environment if rotation fails', () => {
+  it.concurrent('should fall back to user key in hosted environment if rotation fails', () => {
     isHostedSpy.mockReturnValue(true)
 
     module.require = vi.fn(() => {
       throw new Error('Rotation failed')
     })
 
-    // Use gpt-4o which IS in the hosted models list
     const key = getApiKey('openai', 'gpt-4o', 'user-fallback-key')
     expect(key).toBe('user-fallback-key')
   })
 
-  it('should throw error in hosted environment if rotation fails and no user key', () => {
-    isHostedSpy.mockReturnValue(true)
+  it.concurrent(
+    'should throw error in hosted environment if rotation fails and no user key',
+    () => {
+      isHostedSpy.mockReturnValue(true)
 
-    module.require = vi.fn(() => {
-      throw new Error('Rotation failed')
-    })
+      module.require = vi.fn(() => {
+        throw new Error('Rotation failed')
+      })
 
-    // Use gpt-4o which IS in the hosted models list
-    expect(() => getApiKey('openai', 'gpt-4o')).toThrow('No API key available for openai gpt-4o')
+      expect(() => getApiKey('openai', 'gpt-4o')).toThrow('No API key available for openai gpt-4o')
+    }
+  )
+
+  it.concurrent(
+    'should require user key for non-OpenAI/Anthropic providers even in hosted environment',
+    () => {
+      isHostedSpy.mockReturnValue(true)
+
+      const key = getApiKey('other-provider', 'some-model', 'user-key')
+      expect(key).toBe('user-key')
+
+      expect(() => getApiKey('other-provider', 'some-model')).toThrow(
+        'API key is required for other-provider some-model'
+      )
+    }
+  )
+
+  it.concurrent(
+    'should require user key for models NOT in hosted list even if provider matches',
+    () => {
+      isHostedSpy.mockReturnValue(true)
+
+      const key1 = getApiKey('anthropic', 'claude-sonnet-4-20250514', 'user-key-anthropic')
+      expect(key1).toBe('user-key-anthropic')
+
+      expect(() => getApiKey('anthropic', 'claude-sonnet-4-20250514')).toThrow(
+        'API key is required for anthropic claude-sonnet-4-20250514'
+      )
+
+      const key2 = getApiKey('openai', 'gpt-4o-2024-08-06', 'user-key-openai')
+      expect(key2).toBe('user-key-openai')
+
+      expect(() => getApiKey('openai', 'gpt-4o-2024-08-06')).toThrow(
+        'API key is required for openai gpt-4o-2024-08-06'
+      )
+    }
+  )
+
+  it.concurrent('should return empty for ollama provider without requiring API key', () => {
+    isHostedSpy.mockReturnValue(false)
+
+    const key = getApiKey('ollama', 'llama2')
+    expect(key).toBe('empty')
+
+    const key2 = getApiKey('ollama', 'codellama', 'user-key')
+    expect(key2).toBe('empty')
   })
 
-  it('should require user key for non-OpenAI/Anthropic providers even in hosted environment', () => {
-    isHostedSpy.mockReturnValue(true)
+  it.concurrent(
+    'should return empty or user-provided key for vllm provider without requiring API key',
+    () => {
+      isHostedSpy.mockReturnValue(false)
 
-    const key = getApiKey('other-provider', 'some-model', 'user-key')
-    expect(key).toBe('user-key')
+      const key = getApiKey('vllm', 'vllm/qwen-3')
+      expect(key).toBe('empty')
 
-    expect(() => getApiKey('other-provider', 'some-model')).toThrow(
-      'API key is required for other-provider some-model'
-    )
-  })
-
-  it('should require user key for models NOT in hosted list even if provider matches', () => {
-    isHostedSpy.mockReturnValue(true)
-
-    // Models with version suffixes that are NOT in the hosted list should require user API key
-    // even though they're from anthropic/openai providers
-
-    // User provides their own key - should work
-    const key1 = getApiKey('anthropic', 'claude-sonnet-4-20250514', 'user-key-anthropic')
-    expect(key1).toBe('user-key-anthropic')
-
-    // No user key - should throw, NOT use server key
-    expect(() => getApiKey('anthropic', 'claude-sonnet-4-20250514')).toThrow(
-      'API key is required for anthropic claude-sonnet-4-20250514'
-    )
-
-    // Same for OpenAI versioned models not in list
-    const key2 = getApiKey('openai', 'gpt-4o-2024-08-06', 'user-key-openai')
-    expect(key2).toBe('user-key-openai')
-
-    expect(() => getApiKey('openai', 'gpt-4o-2024-08-06')).toThrow(
-      'API key is required for openai gpt-4o-2024-08-06'
-    )
-  })
+      const key2 = getApiKey('vllm', 'vllm/llama', 'user-key')
+      expect(key2).toBe('user-key')
+    }
+  )
 })
 
 describe('Model Capabilities', () => {
@@ -202,7 +227,6 @@ describe('Model Capabilities', () => {
     it.concurrent(
       'should inherit temperature support from provider for dynamically fetched models',
       () => {
-        // OpenRouter models should inherit temperature support from provider capabilities
         expect(supportsTemperature('openrouter/anthropic/claude-3.5-sonnet')).toBe(true)
         expect(supportsTemperature('openrouter/openai/gpt-4')).toBe(true)
       }

--- a/apps/sim/providers/utils.ts
+++ b/apps/sim/providers/utils.ts
@@ -630,11 +630,17 @@ export function getApiKey(provider: string, model: string, userProvidedKey?: str
   // If user provided a key, use it as a fallback
   const hasUserKey = !!userProvidedKey
 
-  // Ollama models don't require API keys - they run locally
+  // Ollama and vLLM models don't require API keys
   const isOllamaModel =
     provider === 'ollama' || useProvidersStore.getState().providers.ollama.models.includes(model)
   if (isOllamaModel) {
     return 'empty' // Ollama uses 'empty' as a placeholder API key
+  }
+
+  const isVllmModel =
+    provider === 'vllm' || useProvidersStore.getState().providers.vllm.models.includes(model)
+  if (isVllmModel) {
+    return userProvidedKey || 'empty' // vLLM uses 'empty' as a placeholder if no key provided
   }
 
   // Use server key rotation for all OpenAI models, Anthropic's Claude models, and Google's Gemini models on the hosted platform

--- a/apps/sim/providers/vllm/index.ts
+++ b/apps/sim/providers/vllm/index.ts
@@ -79,7 +79,15 @@ export const vllmProvider: ProviderConfig = {
     }
 
     try {
-      const response = await fetch(`${baseUrl}/v1/models`)
+      const headers: Record<string, string> = {
+        'Content-Type': 'application/json',
+      }
+
+      if (env.VLLM_API_KEY) {
+        headers.Authorization = `Bearer ${env.VLLM_API_KEY}`
+      }
+
+      const response = await fetch(`${baseUrl}/v1/models`, { headers })
       if (!response.ok) {
         useProvidersStore.getState().setProviderModels('vllm', [])
         logger.warn('vLLM service is not available. The provider will be disabled.')

